src/borrow_check/region_inference/placeholders_and_universes.md

```diff
@@ -0,0 +1,441 @@
+for this section. This variable will have a universe, so let's call
+those universes `U(V1)` and `U(V2)` respectively. (Actually, the only
+one we are going to care about is `U(V1)`.)
+
+When we encounter this constraint, the ordinary procedure is to start
+a DFS from `P`. We keep walking so long as the nodes we are walking
+are present in `value(V2)` and we add those nodes to `value(V1)`. If
+we reach a return point, we add in any `end(X)` elements. That part
+remains unchanged.
+
+But then *after that* we want to iterate over the placeholder `placeholder(x)`
+elements in V2 (each of those must be visible to `U(V2)`, but we
+should be able to just assume that is true, we don't have to check
+it). We have to ensure that `value(V1)` outlives each of those
+placeholder elements.
+
+Now there are two ways that could happen. First, if `U(V1)` can see
+the universe `x` (i.e., `x <= U(V1)`), then we can just add `placeholder(x)`
+to `value(V1)` and be done. But if not, then we have to approximate:
+we may not know what set of elements `placeholder(x)` represents, but we
+should be able to compute some sort of **upper bound** B for it –
+some region B that outlives `placeholder(x)`. For now, we'll just use
+`'static` for that (since it outlives everything) – in the future, we
+can sometimes be smarter here (and in fact we have code for doing this
+already in other contexts). Moreover, since `'static` is in the root
+universe U0, we know that all variables can see it – so basically if
+we find that `value(V2)` contains `placeholder(x)` for some universe `x`
+that `V1` can't see, then we force `V1` to `'static`.
+
+## Extending the "universal regions" check
+
+After all constraints have been propagated, the NLL region inference
+has one final check, where it goes over the values that wound up being
+computed for each universal region and checks that they did not get
+'too large'. In our case, we will go through each placeholder region
+and check that it contains *only* the `placeholder(u)` element it is known to
+outlive. (Later, we might be able to know that there are relationships
+between two placeholder regions and take those into account, as we do
+for universal regions from the fn signature.)
+
+Put another way, the "universal regions" check can be considered to be
+checking constraints like:
+
+```text
+{placeholder(1)}: V1
+```
+
+where `{placeholder(1)}` is like a constant set, and V1 is the variable we
+made to represent the `!1` region.
+
+## Back to our example
+
+OK, so far so good. Now let's walk through what would happen with our
+first example:
+
+```text
+fn(&'static u32) <: fn(&'!1 u32) @ P  // this point P is not imp't here
+```
+
+The region inference engine will create a region element domain like this:
+
+```text
+{ CFG; end('static); placeholder(1) }
+    ---  ------------  ------- from the universe `!1`
+    |    'static is always in scope
+    all points in the CFG; not especially relevant here
```

Comment:
It looks like you inserted some extra spaces here? The underlines don't line up.

Reply:
curious. fixed.